### ISSUE TITLE
Source replacement in cargo config is ignored

### ISSUE THREAD

**Issue by nc7s** [rust-lang/rust-analyzer#16408]
crates.io was down when I met this problem. I use a source replacement so didn't notice it with `cargo add`, but r-a the extension failed to load the workspace:

```
[ERROR rust_analyzer::main_loop] FetchWorkspaceError:
rust-analyzer failed to load workspace: Failed to load the project at /Users/i/Works/salsa/threadlistic/Cargo.toml: Failed to read Cargo metadata from Cargo.toml file /Users/i/Works/salsa/threadlistic/Cargo.toml, Some(Version { major: 1, minor: 75, patch: 0 }): Failed to run `cd "/Users/i/Works/salsa/threadlistic" && "cargo" "metadata" "--format-version" "1" "--manifest-path" "/Users/i/Works/salsa/threadlistic/Cargo.toml" "--filter-platform" "aarch64-apple-darwin"`: `cargo metadata` exited with an error:  Downloading crates ...
warning: spurious network error (3 tries remaining): [35] SSL connect error (LibreSSL SSL_connect: SSL_ERROR_SYSCALL in connection to crates.io:443 )
warning: spurious network error (2 tries remaining): [35] SSL connect error (LibreSSL SSL_connect: SSL_ERROR_SYSCALL in connection to crates.io:443 )
warning: spurious network error (1 tries remaining): [35] SSL connect error (LibreSSL SSL_connect: SSL_ERROR_SYSCALL in connection to crates.io:443 )
error: failed to download from `https://crates.io/api/v1/crates/roxmltree/0.19.0/download`

Caused by:
  [35] SSL connect error (LibreSSL SSL_connect: SSL_ERROR_SYSCALL in connection to crates.io:443 )
```

"Downness" was checked with curl.

**rust-analyzer version**: v0.3.1807

**rustc version**: rustc 1.75.0 (82e1608df 2023-12-21)

**relevant settings**: `~/.cargo/config.toml`

```toml
[source.crates-io]
replace-with = 'tuna'

[source.tuna]
registry = 'sparse+https://mirrors.tuna.tsinghua.edu.cn/crates.io-index/'
```

**Comment by davidbarsky** [comment]
This is an issue with Cargo, not rust-analyzer (but maybe rust-analyzer should surface this error more prominently). What is the behavior you're expecting?

**Comment by nc7s** [comment]
Hmm, weird. I would of course expect it to respect the source replacement, but if it only runs the cargo executable to do flycheck, there's no reason cargo doesn't respect it?

**Comment by nc7s** [comment]
To test that I changed the replacement to an invalid target, and r-a did fail, logging exactly the invalid target. Probably should report this to cargo.
